### PR TITLE
✨ Add controllers tests for real Reconcile loop running with setup-envtest

### LINKED ISSUE #146: Add controllers tests for real Reconcile loop running with setup-envtest
**User Story**

As a developer I would like to run real controller reconcile loop for better testing, i.e. update or full lifecycle

**Detailed Description**

It would be nice to run controller test with `SetupWithManager` in addition to the single `Reconcile(...)` function just one time.
This is a big limitation, as setup-envetest fake Kubernetes cluster already bootstrapped in unit tests but not utilized in integration.
No CR updates can be tested with single `Reconcile` approach, as well as some features, introduced on the Kubernetes API server level, i.e. metadata Observed value, ObjectRefrence, Finalizers, etc. are not allowed.

**Anything else you would like to add:**

It would be nice to add real reconcile integration tests inside the `controller` package in addition to the single `Reconcile` unit tests and before real e2e tests introduced in #147.

/kind feature

### PR DESCRIPTION
**Pull request by dmvolod**
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #146

### CODE REVIEW

**Comment by Jont828** [comment]
/ok-to-test

Still working on e2e tests (like in #163) so please be patient!

File: controllers/helmreleaseproxy/helmreleaseproxy_controller.go [hunk 1]
```diff
@@ -113,7 +115,7 @@ func (r *HelmReleaseProxyReconciler) Reconcile(ctx context.Context, req ctrl.Req
 		return ctrl.Result{}, errors.Wrapf(err, "failed to init patch helper")
 	}
 
-	initalizeConditions(ctx, patchHelper, helmReleaseProxy)
+	initializeConditions(ctx, patchHelper, helmReleaseProxy)
```
Comment (Jont828) [inline controllers/helmreleaseproxy/helmreleaseproxy_controller.go:118]: Nice catch!

File: controllers/controllers_test.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+			}
+
+			return true
+		}, timeout, interval).Should(BeTrue())
+
+		waitForHelmReleaseProxyCondition(client.ObjectKeyFromObject(defaultProxy), func(helmReleaseProxyList []addonsv1alpha1.HelmReleaseProxy) bool {
+			return len(helmReleaseProxyList) == 0
+		})
+	})
+})
```
Comment (Jont828) [inline controllers/controllers_test.go:170]: Newline at EOF

File: controllers/testdata/crd/cluster.x-k8s.io_clusters.yaml [hunk 3]
```diff
@@ -0,0 +1,1386 @@
+---
```
Comment (Jont828) [inline controllers/testdata/crd/cluster.x-k8s.io_clusters.yaml:1]: Now that I've added e2e tests, we have a `test/` folder on the project root. Maybe we could move this file to `test/controllers/data` to avoid putting test data in the controller package.

File: Makefile [hunk 4]
```diff
@@ -264,6 +264,11 @@ generate-modules: ## Run go mod tidy to ensure modules are up to date
 	go mod tidy
 	cd $(TOOLS_DIR); go mod tidy
 
+.PHONY: download-cluster-api-crd
+download-cluster-api-crd: generate-modules ## Run to download Cluster API CRDs for tests
+	cp -r $(shell go env GOPATH)/pkg/mod/sigs.k8s.io/cluster-api@$(shell go list -f "{{.Version}}" -m sigs.k8s.io/cluster-api)/config/crd/bases/cluster.x-k8s.io_clusters.yaml ./controllers/testdata/crd
```
Comment (Jont828) [inline Makefile:269]: Should we add a CAPI version pin in the Makefile since the CAPI version is already pinned in go.mod?

File: Makefile [hunk 5]
```diff
@@ -375,7 +380,7 @@ ARTIFACTS ?= ${ROOT_DIR}/_artifacts
 KUBEBUILDER_ASSETS ?= $(shell $(SETUP_ENVTEST) use --use-env -p path $(KUBEBUILDER_ENVTEST_KUBERNETES_VERSION))
 
 .PHONY: test
-test: $(SETUP_ENVTEST) ## Run unit and integration tests
+test: $(SETUP_ENVTEST) download-cluster-api-crd ## Run unit and integration tests
```
Comment (Jont828) [inline Makefile:383]: I'm not sure about adding this as required dependency for test since ideally you could run unit tests offline.

**Review by Jont828 — COMMENTED** [review]
Nice contribution! Here's a first pass with comments, will circle back for a more thorough review.

File: Makefile (hunk 5 above)
Comment (dmvolod) [reply Makefile:383]: Yes, good point. Will fix it.

File: Makefile (hunk 4 above)
Comment (dmvolod) [reply Makefile:269]: Yes, originally I used this approach (CLUSTER_API_VERSION variable in Makefile), but current approach avoids to keep it and changing every time Clusrer API changing in go.mod

File: internal/kubeconfig.go [hunk 6]
```diff
@@ -25,25 +25,25 @@ import (
...
 type Getter interface {
-	GetClusterKubeconfig(ctx context.Context, cluster *clusterv1.Cluster) (string, error)
+	GetClusterKubeConfig(ctx context.Context, clusterKey ctrlclient.ObjectKey) (string, error)
 }
 
 type KubeconfigGetter struct{}
 
-// GetClusterKubeconfig returns the kubeconfig for a selected Cluster as a string.
-func (k *KubeconfigGetter) GetClusterKubeconfig(ctx context.Context, cluster *clusterv1.Cluster) (string, error) {
+// GetClusterKubeConfig returns the kubeconfig for a selected Cluster as a string.
```
[the remark this answers is not in the document]
Comment (dmvolod) [reply internal/kubeconfig.go:41]: Yes and no, as some projects utilizing this name, but, for example kubectl, kubernetes, etc use CamelCase for it. But both of them can be used.

https://github.com/kubernetes/kubernetes/blob/dbd3f3564ac6cca9a152a3244ab96257e5a4f00c/pkg/scheduler/framework/runtime/framework.go#L160

File: controllers/controllers_test.go (hunk 2 above)
Comment (dmvolod) [reply controllers/controllers_test.go:170]: Good point, will fix it.

File: controllers/testdata/crd/cluster.x-k8s.io_clusters.yaml (hunk 3 above)
Comment (dmvolod) [reply controllers/testdata/crd/cluster.x-k8s.io_clusters.yaml:1]: Yes, sure, as setup-envtest is very flexible, we can move these files outside the controller data.

File: Makefile (hunk 4 above)
Comment (dmvolod) [reply Makefile]: Fix it with introducing `CLUSTER_API_VERSION := $(call get_go_version,sigs.k8s.io/cluster-api)` variable. Also tune `get_go_version` a bit without `awk` utilization and native golang format approach.

File: Makefile (hunk 5 above)
Comment (dmvolod) [reply Makefile]: Fixed

File: controllers/testdata/crd/cluster.x-k8s.io_clusters.yaml (hunk 3 above)
Comment (dmvolod) [reply controllers/testdata/crd/cluster.x-k8s.io_clusters.yaml]: Folder moved

File: controllers/controllers_test.go (hunk 2 above)
Comment (dmvolod) [reply controllers/controllers_test.go:170]: Sorry, not found this catch.

File: internal/kubeconfig.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Jont828) [reply internal/kubeconfig.go:41]: Gotcha, I think we should keep it non-camel cased to follow the Cluster API naming convention since this is a subproject of CAPI if you don't mind.

File: controllers/helmreleaseproxy/helmreleaseproxy_controller_test.go [hunk 7]
```diff
@@ -438,3 +444,9 @@ func TestReconcileDelete(t *testing.T) {
 		})
 	}
 }
+
+func init() {
```
Comment (Jont828) [inline controllers/helmreleaseproxy/helmreleaseproxy_controller_test.go:448]: Why is this needed in for the unit tests in this file?

File: controllers/suite_test.go [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+		Expect(err).ToNot(HaveOccurred(), "failed to run manager")
+	}()
+})
+
+var _ = AfterSuite(func() {
+	cancel()
+	By("tearing down the test environment")
+	err := testEnv.Stop()
+	Expect(err).NotTo(HaveOccurred())
+})
```
Comment (Jont828) [inline controllers/suite_test.go:164]:
```suggestion
})

```

File: internal/kubeconfig.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (dmvolod) [reply internal/kubeconfig.go:41]: Yes, good point. Let's keep it as is. Will return it back in this PR.

File: controllers/helmreleaseproxy/helmreleaseproxy_controller_test.go (hunk 7 above)
Comment (dmvolod) [reply controllers/helmreleaseproxy/helmreleaseproxy_controller_test.go:448]: As unit-tests utilizing`fake.NewClientBuilder()` it requires sheme's to be regitered.

**Comment by Jont828** [comment]
Thanks for the contribution!

/lgtm
/approve
